File: backend/app/models/event.py
from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime
from typing import Optional


class EventCreate(BaseModel):
    name: str = Field(min_length=1)
    slug: str = Field(min_length=1, pattern=r"^[a-z0-9][a-z0-9-]*[a-z0-9]$")
    vm_os: str
    vm_distribution: str = ""
    vm_mem: int = Field(gt=0)
    vm_vcpus: int = Field(gt=0)
    vm_disk_size: int = Field(gt=0)
    deadline: datetime
    max_vms: int = Field(gt=0)


class EventUpdate(BaseModel):
    name: Optional[str] = None
    vm_os: Optional[str] = None
    vm_distribution: Optional[str] = None
    vm_mem: Optional[int] = Field(default=None, gt=0)
    vm_vcpus: Optional[int] = Field(default=None, gt=0)
    vm_disk_size: Optional[int] = Field(default=None, gt=0)
    deadline: Optional[datetime] = None
    max_vms: Optional[int] = Field(default=None, gt=0)


class EventParticipantRead(BaseModel):
    id: UUID
    participant_name: str
    vm_id: UUID
    created_at: datetime


class EventRead(BaseModel):
    id: UUID
    slug: str
    name: str
    vm_os: str
    vm_distribution: str
    vm_mem: int
    vm_vcpus: int
    vm_disk_size: int
    deadline: datetime
    max_vms: int
    created_at: datetime
    created_by: str
    participants_count: int = 0
    participants: list[EventParticipantRead] = []


class EventJoinRequest(BaseModel):
    participant_name: str = Field(min_length=1, max_length=64)


class EventJoinResponse(BaseModel):
    vm_id: UUID
    vm_name: str
    credential_name: str
    credential_password: str

File: frontend/app/lib/types/event.ts
import { z } from "zod";

export const EventParticipantSchema = z.object({
  id: z.string().uuid(),
  participant_name: z.string(),
  vm_id: z.string().uuid(),
  created_at: z.string(),
});

export const EventSchema = z.object({
  id: z.string().uuid(),
  slug: z.string(),
  name: z.string(),
  vm_os: z.string(),
  vm_distribution: z.string(),
  vm_mem: z.number(),
  vm_vcpus: z.number(),
  vm_disk_size: z.number(),
  keyboard_layout: z.string().nullable().optional(),
  deadline: z.string().transform((s) => (s.endsWith("Z") ? s : s + "Z")),
  max_vms: z.number(),
  created_at: z.string(),
  created_by: z.string(),
  participants_count: z.number(),
  participants: z.array(EventParticipantSchema),
});

export type Event = z.infer<typeof EventSchema>;
export type EventParticipant = z.infer<typeof EventParticipantSchema>;

export const CreateEventPayloadSchema = z.object({
  name: z.string().min(1),
  slug: z
    .string()
    .min(1)
    .regex(/^[a-z0-9][a-z0-9-]*[a-z0-9]$/),
  vm_os: z.string().min(1),
  vm_distribution: z.string().default(""),
  vm_mem: z.number().positive(),
  vm_vcpus: z.number().positive(),
  vm_disk_size: z.number().positive(),
  keyboard_layout: z.string().nullable().optional(),
  deadline: z.string(),
  max_vms: z.number().positive(),
});

export type CreateEventPayload = z.infer<typeof CreateEventPayloadSchema>;

export const UpdateEventPayloadSchema = z.object({
  name: z.string().min(1).optional(),
  vm_os: z.string().min(1).optional(),
  vm_mem: z.number().positive().optional(),
  vm_vcpus: z.number().positive().optional(),
  vm_disk_size: z.number().positive().optional(),
  deadline: z.string().optional(),
  max_vms: z.number().positive().optional(),
});

export type UpdateEventPayload = z.infer<typeof UpdateEventPayloadSchema>;

export const EventJoinResponseSchema = z.object({
  vm_id: z.string().uuid(),
  vm_name: z.string(),
  credential_name: z.string(),
  credential_password: z.string(),
});

export type EventJoinResponse = z.infer<typeof EventJoinResponseSchema>;
